### PR TITLE
New framework feature: mask output streams with fill value

### PR DESCRIPTION
**Pull request by cbegeman**
This PR adds a new optional feature to fill variables in output streams with the `missing_value` defined in the Registry everywhere a masking array, designated with its registry name by a new registry attribute `missing_value_mask`, is equal to zero. These attributes can be added either to `var`s or to `vararray`s; when added to a `vararray`, masking is applied to each `var` in the `vararray`.

The masking is only applied to variables who (a) have a `missing_value` and `missing_value_mask` attribute and (b) are members of a stream that has the attribute `useMissingValMask='true'`. `useMissingValMask='false'` by default. Thus, changes can be made to Registry attributes without affecting the default behavior of streams.

Masking is supported for Real fields with dimensions 1-5 when the masking array is an integer array with dimensions 1-3 _and_ the shape of the masking array matches up to 3 contiguous dimensions of the Real field. 

The structure of code changes is as follows:

1. `gen_inc.c` adds `missing_value_mask` as specified in the Registry as an attribute of `var`s and `vararray`s and stores the value of `missing_value_mask` in the real field type structure as `maskName`. 
2. `mpas_bootstrap_framework_phase2` calls `mpas_mask_assign_pointer`, which assigns a pointer from the real field structure to the masking array. The name of the pointer is `missingValueMaskXD` where X is the dimension of the masking array. 
3. `xml_stream_parser.c` ensures that `useMissingValMask` is added as an attribute of the stream. `stream_mgr_create_stream_c` converts the string value of `useMissingValMask` given in the streams xml file to the logical value of `useMissingValMask` in the stream type structure.
4. `MPAS_writeStream` applies the masking if the command line argument `useMissingValMask` (which derives from the stream attribute) is `.true.` and the array dimensions are appropriate. It applies the masking through a call  to `mpas_mask_field`.
5. `mpas_field_routines.F` contains `mpas_mask_assign_pointer` and `mpas_mask_field` for multiple mask and field dimension combinations.
6. Support was added for the `time_series_stats` analysis member through changes to `duplicate_field_array.inc` and assigning pointer to `missingValueMaskXd` in subroutines of `mpas_ocn_time_series_stats.F`. The latter needed to be located in dimension-specific routines so that the appropriate X values were used.
7. The `global_ocean` test cases had an implicit assumption that `normalVelocity` would be initialized to 0 that needed to be made explicit in `mpas_ocn_init_global_ocean.F` in order for the PR test suite to pass.

**To use this feature:**

1. If the variable you want to mask does not have `missing_value_mask="maskingArray"` in its entry in the Registry, add this line.
2. Add `useMissingValMask="true"` to the stream entry to which you would like to apply the masking.

### CODE REVIEW

**Comment by cbegeman** [comment]
@akturner and @xylar, In case you have a chance to take a look, I've gotten as far as this line before hitting errors https://github.com/E3SM-Ocean-Discussion/E3SM/blob/569a4d7981936b2a2802b7733ee00241d7d113c1/components/mpas-framework/src/framework/mpas_field_routines.F#L2988
 The shape I get for the `normalVelocity` array to be masked is (1,0) which is clearly incorrect. 

For your reference, this is the line that calls the masking routine:
https://github.com/E3SM-Ocean-Discussion/E3SM/blob/569a4d7981936b2a2802b7733ee00241d7d113c1/components/mpas-framework/src/framework/mpas_io_streams.F#L3656

Note: for now, I've commented out a bunch of stuff that pertains to other field/mask dimension combinations.

I get the `nParticles` error when the code crashes in the masking routine, so that's a red herring I think.

File: components/mpas-framework/src/framework/mpas_field_routines.F [hunk 1]
```diff
@@ -2858,6 +2869,183 @@ subroutine mpas_allocate_mold_1dchar(dst, src)!{{{
...
+      real (kind=RKIND), dimension(:,:), intent(out) :: maskedField !< Input/Output: Field to be masked
+      ! TODO consider whether this should be of field type or 3d real array
+      !type (field3DReal), intent(out) :: maskedField !< Input/Output: Field to be masked
+
+      real (kind=RKIND), dimension(:,:), pointer :: fieldArray
+      character (len=100) :: log_string
+      integer, dimension(:,:), pointer :: maskArray
+
+      call mpas_log_write('Masking array '//trim(field % fieldName))
+      fieldArray = field % array
```
Comment (akturner) [inline components/mpas-framework/src/framework/mpas_field_routines.F:2987]: fieldArray is a pointer that has neither been allocated or assigned to another pointer at this point.
This line may need to be fieldArray => field % array to assign the pointer if that was your intent

Comment (cbegeman) [reply components/mpas-framework/src/framework/mpas_field_routines.F:2987]: Great, thanks!

File: components/mpas-framework/src/framework/mpas_io_streams.F [hunk 2]
```diff
@@ -3641,30 +3644,50 @@ subroutine MPAS_writeStream(stream, frame, ierr)
...
+                        if (ASSOCIATED( field_2dreal_ptr % missingValueMask2D )) then
+                           call mpas_mask_field2d_mask2d_real(field_2dreal_ptr, field_2dreal_ptr % missingValueMask2d, ownedSize, real2d_temp)
+                           call mpas_log_write('.....masking completed')
+                        else
+                           real2d_temp(:,i:i+ownedSize-1) = field_2dreal_ptr % array(:,1:ownedSize)
+                        endif
                      end if
                      i = i + ownedSize
                      if ( .not. stream % blockWrite ) then
+                        call mpas_log_write('Go to next')
```
Comment (cbegeman) [inline components/mpas-framework/src/framework/mpas_io_streams.F:3668]: @akturner and @xylar I don't have any more time to work on this today, but I was able to figure out that the crash happens between this line and the first line of the do-while loop. 

The log is (in part):
```
 Writing out 2d field normalVelocity
 field_2dreal_ptr is normalVelocity
  i,OwnedSize,dimName =            1         150 nEdges
 ... isVar decomposed, assigning real2d_temp
 Masking array normalVelocity
  fieldArray shape=          20         237
  maskedField shape=          20         150
 .....masking completed
 Go to next

Comment (xylar) [reply components/mpas-framework/src/framework/mpas_io_streams.F:3668]: Great, I'll see if I can make any headway this evening.

Comment (xylar) [reply components/mpas-framework/src/framework/mpas_io_streams.F:3668]: I got stuck on my time coordinate branch.  It seemed like I had everything working but then it failed catastrophically with a bunch of useless PIO errors.

So once again I didn't get to this -- I had no debugging patience left.  Maybe we can work on it together tomorrow during the work time.

Comment (xylar) [reply components/mpas-framework/src/framework/mpas_io_streams.F:3668]: I got stuck on my time coordinate branch.  It seemed like I had everything working but then it failed catastrophically with a bunch of useless PIO errors.

So once again I didn't get to this -- I had no debugging patience left.  Maybe we can work on it together tomorrow during the work time.

File: components/mpas-framework/src/framework/mpas_dmpar.F [hunk 3]
```diff
@@ -8832,23 +8832,30 @@ subroutine mpas_dmpar_exch_group_pack_buffer_field3d_real(exchangeGroup, field,
...
              if ( exchListPtr % endPointID == commListPtr % procID ) then
                do iExch = 1, exchListPtr % nList
+                 call mpas_log_write('iter on nList, '//trim(fieldCursor % dimNames(1))//trim(fieldCursor % dimNames(2)))
                  do j = 1, fieldCursor % dimSizes(2)
                    do k = 1, fieldCursor % dimSizes(1)
+                     call mpas_log_write('get iBuffer')
                      iBuffer = (exchListPtr % destList(iExch) - 1) * fieldCursor % dimSizes(1) * fieldCursor % dimSizes(2) &
                                + (j - 1) * fieldCursor % dimSizes(1) + k + bufferOffset
+                     call mpas_log_write('get array')
                      commListPtr % rbuffer(iBuffer) = fieldCursor % array(k, j, exchListPtr % srcList(iExch))
```
Comment (cbegeman) [inline components/mpas-framework/src/framework/mpas_dmpar.F:8849]: @akturner or others, I'm on holiday next week. As I mentioned on slack, the 2d mask on a 2d var appears to be working (at the commit https://github.com/E3SM-Ocean-Discussion/E3SM/pull/25/commits/345440573f7f72a2183efee3e0b60ef3da049603). I added some of the infrastructure for adding a 2d mask to the 3d vararray activeTracers. However, there is a seg fault at this line that I didn't get a chance to track down. Suggestions welcome!

**Comment by cbegeman** [comment]
Note: E3SM-Ocean-Discussion/master needs to be brought up to E3SM-Project/master

**Comment by xylar** [comment]
@cbegeman, this is **amazing** work!  I finally had a chance to look through the code and everything looks good to me.  I think it makes a lot of sense to include the MPAS-Ocean changes that are here even though this is mostly a framework PR.

One think I would suggest is making the identical changes in the MPAS-Seaice time-series-stats code.  I had to do that in #23 as well.  We'll need help from @akturner to test it, of course.

**Comment by xylar** [comment]
@cbegeman, don't forget to rename the branch `mpas-framework/new-fill-value-mask-feature` before you make a PR to E3SM-Project.

**Comment by xylar** [comment]
@cbegeman, I'm seeing a bunch of debug output that should be turned off again before moving to E3SM-Project:
```
 Doing timestep 0001-01-01_00:05:00
 Writing out 2d field normalVelocity
 field_2dreal_ptr is normalVelocity
  i,OwnedSize,dimName =            1         606 nEdges
 ... isVar decomposed, assigning real2d_temp
 Masking array normalVelocity
  fieldArray shape=          20         937
  maskedField shape=          20         606
 .....masking completed
 Go to next
...
```

**Review by xylar — APPROVED** [review]
From the ocean and framework side, I think this is ready to go!  I ran the `nightly` test suite and all tests passed and are bit-for-bit with a baseline from the current compass master and its E3SM-Project submodule.

I'm approving although I think the sea-ice time series stats changes are still needed as I mentioned above.

**Comment by cbegeman** [comment]
@xylar Thank you so much for taking the time to review this!

**Comment by cbegeman** [comment]
> I think it makes a lot of sense to include the MPAS-Ocean changes that are here even though this is mostly a framework PR.

Just to be clear, you think we should also include the changes to the mpas-ocean registry that result in `layerThickness`, `normalVelocity`, and `activeTracers` being masked?

**Comment by cbegeman** [comment]
> One think I would suggest is making the identical changes in the MPAS-Seaice time-series-stats code. I had to do that in #23 as well. We'll need help from @akturner to test it, of course.

@akturner I added this commit [eaae351](https://github.com/E3SM-Ocean-Discussion/E3SM/pull/25/commits/eaae3511d1122d50ae04bd15ed2f4e74d0414831) that ensures the `missing_value_mask` attribute gets copied in generating mpas-seaice time series stats. I think what you need to do to test it is to identify a masking variable that can be used on one of the time series stats variables and add it to the Registry following https://github.com/E3SM-Ocean-Discussion/E3SM/blob/eaae3511d1122d50ae04bd15ed2f4e74d0414831/components/mpas-ocean/src/Registry.xml#L2114

Then you can run a sea ice test case involving time series stats and check that the variable is masked.
Thanks!

**Comment by xylar** [comment]
> @xylar Just to be clear, you think we should also include the changes to the mpas-ocean registry that result in `layerThickness`, `normalVelocity`, and `activeTracers` being masked?

Yeah, I thought about it a bit and I think it makes sense to modify those fields to show that this works the way it's intended.

**Comment by akturner** [comment]
@cbegeman: Apologies for the delay in looking at this. This doesn't compile for the sea ice core producing the following error:
```
mpif90 -D_MPI -DUSE_PIO2 -DCORE_SEAICE -DMPAS_NAMELIST_SUFFIX=seaice -DMPAS_EXE_NAME=seaice_model -DMPAS_NATIVE_TIMERS -DMPAS_GIT_VERSION=v2.0.0-1713-g8fba8ab530-dirty -O3 -m64 -ffree-line-length-none -fconvert=big-endian -ffree-form -ffpe-summary=none  -fdefault-real-8 -fdefault-double-8 -c mpas_seaice_time_series_stats.F -I../../../mpas-framework/src/external/esmf_time_f90 -I../../../mpas-framework/src/framework -I../../../mpas-framework/src/operators -I../column -I../shared -I../model_forward -I/home/akt/Work/libraries/scorpio/scorpio-v1.1.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/libraries/netcdf-c/netcdf-c-4.7.4/gcc/7.5/include -I/home/akt/Work/libraries/pnetcdf/pnetcdf-1.12.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/libraries/scorpio/scorpio-v1.1.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/libraries/netcdf-c/netcdf-c-4.7.4/gcc/7.5/include -I/home/akt/Work/libraries/netcdf-fortran/netcdf-fortran-4.5.3/gcc/7.5/include -I/home/akt/Work/libraries/pnetcdf/pnetcdf-1.12.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/column -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/shared -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/analysis_members -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/model_forward
mpas_seaice_time_series_stats.F:2036:2:

   dst % missingValueMask1D => src % missingValueMask2D
  1
Error: Different types in pointer assignment at (1); attempted assignment of TYPE(field2dinteger) to TYPE(field1dinteger)
mpas_seaice_time_series_stats.F:2037:2:

   dst % missingValueMask1D => src % missingValueMask3D
  1
Error: Different types in pointer assignment at (1); attempted assignment of TYPE(field3dinteger) to TYPE(field1dinteger)
mpas_seaice_time_series_stats.F:2027:2:

   dst % missingValueMask1D => src % missingValueMask2D
  1
Error: Different types in pointer assignment at (1); attempted assignment of TYPE(field2dinteger) to TYPE(field1dinteger)
mpas_seaice_time_series_stats.F:2028:2:

   dst % missingValueMask1D => src % missingValueMask3D
  1
Error: Different types in pointer assignment at (1); attempted assignment of TYPE(field3dinteger) to TYPE(field1dinteger)
mpas_seaice_time_series_stats.F:2018:2:

   dst % missingValueMask1D => src % missingValueMask2D
  1
Error: Different types in pointer assignment at (1); attempted assignment of TYPE(field2dinteger) to TYPE(field1dinteger)
mpas_seaice_time_series_stats.F:2019:2:

   dst % missingValueMask1D => src % missingValueMask3D
  1
Error: Different types in pointer assignment at (1); attempted assignment of TYPE(field3dinteger) to TYPE(field1dinteger)
mpas_seaice_time_series_stats.F:2010:2:

   dst % missingValueMask1D => src % missingValueMask2D
  1
Error: Different types in pointer assignment at (1); attempted assignment of TYPE(field2dinteger) to TYPE(field1dinteger)
Makefile:21: recipe for target 'mpas_seaice_time_series_stats.o' failed
```
Also the following warnings were generated during the build:
```
mpif90 -D_MPI -DUSE_PIO2 -DCORE_SEAICE -DMPAS_NAMELIST_SUFFIX=seaice -DMPAS_EXE_NAME=seaice_model -DMPAS_NATIVE_TIMERS -DMPAS_GIT_VERSION=v2.0.0-1713-g8fba8ab530-dirty -O3 -m64 -ffree-line-length-none -fconvert=big-endian -ffree-form -ffpe-summary=none  -fdefault-real-8 -fdefault-double-8 -c mpas_io_streams.F -I/home/akt/Work/libraries/scorpio/scorpio-v1.1.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/libraries/netcdf-c/netcdf-c-4.7.4/gcc/7.5/include -I/home/akt/Work/libraries/pnetcdf/pnetcdf-1.12.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/libraries/scorpio/scorpio-v1.1.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/libraries/netcdf-c/netcdf-c-4.7.4/gcc/7.5/include -I/home/akt/Work/libraries/netcdf-fortran/netcdf-fortran-4.5.3/gcc/7.5/include -I/home/akt/Work/libraries/pnetcdf/pnetcdf-1.12.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/column -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/shared -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/analysis_members -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/model_forward -I../external/esmf_time_f90
mpas_io_streams.F:4086:95:

                                                          field_cursor % real3dField % dimSizes(4), &
                                                                                               1
Warning: Array reference at (1) is out of bounds (4 > 3) in dimension 1
mpas_io_streams.F:4097:95:

                                                          field_cursor % real3dField % dimSizes(4), &
                                                                                               1
Warning: Array reference at (1) is out of bounds (4 > 3) in dimension 1
mpas_io_streams.F:4108:95:

                                                          field_cursor % real3dField % dimSizes(4), &
                                                                                               1
Warning: Array reference at (1) is out of bounds (4 > 3) in dimension 1
```

**Comment by cbegeman** [comment]
@akturner I see the problem. Let me push the fix real quick. Downside of me not knowing how to build mpas-seaice is not catching these things

**Comment by akturner** [comment]
@cbegeman: Build still generates new warning messages:
```
mpif90 -D_MPI -DUSE_PIO2 -DCORE_SEAICE -DMPAS_NAMELIST_SUFFIX=seaice -DMPAS_EXE_NAME=seaice_model -DMPAS_NATIVE_TIMERS -DMPAS_GIT_VERSION=v2.0.0-1713-g8fba8ab530-dirty -O3 -m64 -ffree-line-length-none -fconvert=big-endian -ffree-form -ffpe-summary=none  -fdefault-real-8 -fdefault-double-8 -c mpas_io_streams.F -I/home/akt/Work/libraries/scorpio/scorpio-v1.1.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/libraries/netcdf-c/netcdf-c-4.7.4/gcc/7.5/include -I/home/akt/Work/libraries/pnetcdf/pnetcdf-1.12.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/libraries/scorpio/scorpio-v1.1.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/libraries/netcdf-c/netcdf-c-4.7.4/gcc/7.5/include -I/home/akt/Work/libraries/netcdf-fortran/netcdf-fortran-4.5.3/gcc/7.5/include -I/home/akt/Work/libraries/pnetcdf/pnetcdf-1.12.1/gcc/7.5/openmpi/2.1.1/include -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/column -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/shared -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/analysis_members -I/home/akt/Work/MPAS-Seaice/satisfy_prs/pr_cbegeman_new-fill-value-mask-feature/development/E3SM/components/mpas-seaice/src/model_forward -I../external/esmf_time_f90
mpas_io_streams.F:4086:95:

                                                          field_cursor % real3dField % dimSizes(4), &
                                                                                               1
Warning: Array reference at (1) is out of bounds (4 > 3) in dimension 1
mpas_io_streams.F:4097:95:

                                                          field_cursor % real3dField % dimSizes(4), &
                                                                                               1
Warning: Array reference at (1) is out of bounds (4 > 3) in dimension 1
mpas_io_streams.F:4108:95:

                                                          field_cursor % real3dField % dimSizes(4), &
                                                                                               1
Warning: Array reference at (1) is out of bounds (4 > 3) in dimension 1
```

**Comment by cbegeman** [comment]
@akturner Thanks for catching that! I have now fixed up the pointers so that the appropriate dimensions are used for 4D and 5D fields. Let me know if that fixes it for you.

**Comment by cbegeman** [comment]
@akturner and @xylar Thank you both so much for reviewing and testing!

**Comment by xylar** [comment]
@cbegeman, I'm closing this PR now that https://github.com/E3SM-Project/E3SM/pull/5154 is open.  I hope you don't mind.

**Pull request closed without merging**